io/shared/src/main/scala/scalaz/effect/IO.scala

```diff
@@ -178,29 +167,33 @@ sealed abstract class IO[E, A] { self =>
    * The error parameter of the returned `IO` may be chosen arbitrarily, since
    * it is guaranteed the `IO` action does not raise any errors.
    */
-  final def attempt[E2]: IO[E2, E \/ A] = (self.tag: @switch) match {
-    case IO.Tags.Point =>
-      val io = self.asInstanceOf[IO.Point[E, A]]
+  final def attempt[E2, B](err: E => B)(succ: A => B): IO[E2, B] =
```

Comment:
This is not a good signature for `attempt` and will involve huge overhead (as you can see in the need to call `flatAttempt`).

Instead, I recommend the previously suggested signature:

```scala
final def attempt[E2, B](err: E => IO[E2, B], succ: A => IO[E2, B]): IO[E2, B]
```

Then just capture these `err` and `succ` closures inside the `Attempt` constructor and deal with them in `RTS`.

Reply:
We can just store these `err` and `succ` closures into a data structure by modifying the signature of the `Attempt` class. So no flat mapping is required.

The user can always recapture this old version of `attempt` from the new one, but even your code changes here suggest "flatAttempt" (the new `attempt`) will be the default.

Reply:
> Then just capture these err and succ closures inside the Attempt constructor and deal with them in RTS.

Ok so we replace the whole `match` with just a wrapping with `Attempt`, I was gonna add the closures anyway

Reply:
Yes, so it becomes very simple, also the RTS will "like" to work with this representation, although it does involve some non-obvious changes that I will help with...